ouroboros-network/src/Data/Sequence/Strict.hs

```diff
@@ -0,0 +1,221 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE ViewPatterns #-}
+
+-- | Strict variants of 'Seq' operations.
+--
+-- TODO: Move this to @cardano-prelude@ as a new module, @Data.Sequence.Strict@.
+module Data.Sequence.Strict
+  ( StrictSeq (.., Empty, (:<|), (:|>))
+
+    -- * Construction
+    , empty
+    , singleton
+    , (<|)
+    , (|>)
+    , (><)
+    , fromList
+
+    -- * Deconstruction
+    -- | Additional functions for deconstructing sequences are available
+    -- via the 'Foldable' instance of 'Seq'.
+
+    -- ** Queries
+    , null
+    , length
+
+    -- * Scans
+    , scanl
+
+    -- * Sublists
+
+    -- ** Sequential searches
+    , dropWhileL
+    , dropWhileR
+    , spanl
+
+    -- * Indexing
+    , take
+    , drop
+    , splitAt
+  ) where
+
+import           Prelude hiding (drop, length, null, scanl, splitAt, take)
+import           Cardano.Prelude (NoUnexpectedThunks (..), forceElemsToWHNF,
+                     noUnexpectedThunksInValues)
+
+import           Codec.Serialise (Serialise)
+import           Data.Foldable (foldl', toList)
+import           Data.Sequence (Seq)
+import qualified Data.Sequence as Seq
+
+infixr 5 ><
+infixr 5 <|
+infixl 5 |>
+
+infixr 5 :<|
+infixl 5 :|>
+
+-- | A @newtype@ wrapper around a 'Seq', representing a general-purpose finite
+-- sequence that is strict in its values.
+--
+-- This strictness is not enforced at the type level, but rather by the
+-- construction functions in this module. These functions essentially just
+-- wrap the original "Data.Sequence" functions while forcing the provided
+-- value to WHNF.
+newtype StrictSeq a = StrictSeq (Seq a)
+  deriving (Eq, Ord, Show)
+
+deriving instance Foldable StrictSeq
+deriving instance Semigroup (StrictSeq a)
+deriving instance Serialise a => Serialise (StrictSeq a)
```

Comment:
I think @edsko would like to see `deriving newtype instance` here :slightly_smiling_face:

Reply:
I went ahead and used the `deriving newtype (Foldable, Semigroup, Serialise)` syntax.